File: api/v1/schemas/admin_mission_control.py
# -*- coding: utf-8 -*-
"""Schemas for the read-only admin mission control cockpit."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class _AdminMissionControlModel(BaseModel):
    model_config = ConfigDict(populate_by_name=True)


class AdminMissionControlPosture(_AdminMissionControlModel):
    landed_foundation: bool = Field(alias="landedFoundation")
    evidence_tooling_exists: bool = Field(alias="evidenceToolingExists")
    real_operator_evidence_missing: bool = Field(alias="realOperatorEvidenceMissing")
    approval_required: bool = Field(alias="approvalRequired")
    public_launch_no_go: bool = Field(alias="publicLaunchNoGo")


class AdminMissionControlSourceRef(_AdminMissionControlModel):
    kind: Literal["admin_route", "api", "doc", "test", "script", "fixture"]
    label: str
    ref: str


class AdminMissionControlDomainSlice(_AdminMissionControlModel):
    id: str
    title: str
    status: str
    status_label: str = Field(alias="statusLabel")
    summary: str
    posture: AdminMissionControlPosture
    read_only: bool = Field(default=True, alias="readOnly")
    no_external_calls: bool = Field(default=True, alias="noExternalCalls")
    live_enforcement: bool = Field(default=False, alias="liveEnforcement")
    runtime_behavior_changed: bool = Field(default=False, alias="runtimeBehaviorChanged")
    data_sources: List[str] = Field(default_factory=list, alias="dataSources")
    evidence_refs: List[AdminMissionControlSourceRef] = Field(default_factory=list, alias="evidenceRefs")
    blocker_refs: List[AdminMissionControlSourceRef] = Field(default_factory=list, alias="blockerRefs")
    approval_refs: List[AdminMissionControlSourceRef] = Field(default_factory=list, alias="approvalRefs")
    linked_admin_routes: List[str] = Field(default_factory=list, alias="linkedAdminRoutes")
    ops_status: Optional[Dict[str, Any]] = Field(default=None, alias="opsStatus")
    limitations: List[str] = Field(default_factory=list)


class AdminMissionControlSummary(_AdminMissionControlModel):
    domain_count: int = Field(alias="domainCount")
    landed_foundation_count: int = Field(alias="landedFoundationCount")
    evidence_tooling_count: int = Field(alias="evidenceToolingCount")
    real_operator_evidence_missing_count: int = Field(alias="realOperatorEvidenceMissingCount")
    approval_required_count: int = Field(alias="approvalRequiredCount")
    public_launch_no_go_count: int = Field(alias="publicLaunchNoGoCount")


class AdminMissionControlResponse(_AdminMissionControlModel):
    generated_at: str = Field(alias="generatedAt")
    read_only: bool = Field(default=True, alias="readOnly")
    no_external_calls: bool = Field(default=True, alias="noExternalCalls")
    live_enforcement: bool = Field(default=False, alias="liveEnforcement")
    runtime_behavior_changed: bool = Field(default=False, alias="runtimeBehaviorChanged")
    public_launch_approved: bool = Field(default=False, alias="publicLaunchApproved")
    release_approved: bool = Field(default=False, alias="releaseApproved")
    launch_verdict: Literal["NO_GO"] = Field(default="NO_GO", alias="launchVerdict")
    ops_snapshot_available: bool = Field(alias="opsSnapshotAvailable")
    summary: AdminMissionControlSummary
    domains: List[AdminMissionControlDomainSlice]
    posture_legend: Dict[str, str] = Field(alias="postureLegend")
    metadata: Dict[str, Any] = Field(default_factory=dict)

File: apps/dsa-web/src/api/adminMissionControl.ts
import apiClient from './index';
import { toCamelCase } from './utils';

export interface MissionControlPosture {
  landedFoundation: boolean;
  evidenceToolingExists: boolean;
  realOperatorEvidenceMissing: boolean;
  approvalRequired: boolean;
  publicLaunchNoGo: boolean;
}

export interface MissionControlSourceRef {
  kind: 'admin_route' | 'api' | 'doc' | 'test' | 'script' | 'fixture' | string;
  label: string;
  ref: string;
}

export interface MissionControlDomainSlice {
  id: string;
  title: string;
  status: string;
  statusLabel: string;
  summary: string;
  posture: MissionControlPosture;
  readOnly: boolean;
  noExternalCalls: boolean;
  liveEnforcement: boolean;
  runtimeBehaviorChanged: boolean;
  dataSources: string[];
  evidenceRefs: MissionControlSourceRef[];
  blockerRefs: MissionControlSourceRef[];
  approvalRefs: MissionControlSourceRef[];
  linkedAdminRoutes: string[];
  opsStatus?: {
    available?: boolean;
    status?: string;
    reasonCode?: string | null;
    summary?: Record<string, unknown>;
    limitations?: string[];
  } | null;
  limitations: string[];
}

export interface MissionControlSummary {
  domainCount: number;
  landedFoundationCount: number;
  evidenceToolingCount: number;
  realOperatorEvidenceMissingCount: number;
  approvalRequiredCount: number;
  publicLaunchNoGoCount: number;
}

export interface MissionControlResponse {
  generatedAt: string;
  readOnly: boolean;
  noExternalCalls: boolean;
  liveEnforcement: boolean;
  runtimeBehaviorChanged: boolean;
  publicLaunchApproved: boolean;
  releaseApproved: boolean;
  launchVerdict: 'NO_GO' | string;
  opsSnapshotAvailable: boolean;
  summary: MissionControlSummary;
  domains: MissionControlDomainSlice[];
  postureLegend: Record<string, string>;
  metadata: Record<string, unknown>;
}

function safeArray<T>(value: unknown): T[] {
  return Array.isArray(value) ? value as T[] : [];
}

function normalizeDomain(value: Record<string, unknown>): MissionControlDomainSlice {
  const normalized = toCamelCase<MissionControlDomainSlice>(value);
  return {
    ...normalized,
    dataSources: safeArray<string>(normalized.dataSources),
    evidenceRefs: safeArray<MissionControlSourceRef>(normalized.evidenceRefs),
    blockerRefs: safeArray<MissionControlSourceRef>(normalized.blockerRefs),
    approvalRefs: safeArray<MissionControlSourceRef>(normalized.approvalRefs),
    linkedAdminRoutes: safeArray<string>(normalized.linkedAdminRoutes),
    limitations: safeArray<string>(normalized.limitations),
    readOnly: normalized.readOnly === true,
    noExternalCalls: normalized.noExternalCalls === true,
    liveEnforcement: normalized.liveEnforcement === true,
    runtimeBehaviorChanged: normalized.runtimeBehaviorChanged === true,
  };
}

export const adminMissionControlApi = {
  async getSnapshot(): Promise<MissionControlResponse> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/admin/mission-control');
    const normalized = toCamelCase<MissionControlResponse>(response.data);
    return {
      ...normalized,
      readOnly: normalized.readOnly === true,
      noExternalCalls: normalized.noExternalCalls === true,
      liveEnforcement: normalized.liveEnforcement === true,
      runtimeBehaviorChanged: normalized.runtimeBehaviorChanged === true,
      publicLaunchApproved: normalized.publicLaunchApproved === true,
      releaseApproved: normalized.releaseApproved === true,
      opsSnapshotAvailable: normalized.opsSnapshotAvailable === true,
      domains: safeArray<Record<string, unknown>>(normalized.domains).map(normalizeDomain),
      postureLegend: normalized.postureLegend || {},
      metadata: normalized.metadata || {},
    };
  },
};
